### PR TITLE
Implement Entity#setPosition(Vector3d) and fix Entity#setLocation(ServerLocation)

### PR DESCRIPTION
**Pull request by Lignium**

### CODE REVIEW

File: src/mixins/java/org/spongepowered/common/mixin/api/minecraft/world/entity/EntityMixin_API.java [hunk 1]
```diff
@@ -120,6 +119,12 @@ public Vector3d position() {
         return new Vector3d(this.shadow$getX(), this.shadow$getY(), this.shadow$getZ());
     }
 
+    @Override
+    public boolean setPosition(final Vector3d position) {
+        Preconditions.checkNotNull(position, "The position was null!");
```
Comment (Zidane) [inline src/mixins/java/org/spongepowered/common/mixin/api/minecraft/world/entity/EntityMixin_API.java]: Uses `Objects.requireNonNull`

Comment (Zidane) [reply src/mixins/java/org/spongepowered/common/mixin/api/minecraft/world/entity/EntityMixin_API.java]: Also inline it in the return

File: src/mixins/java/org/spongepowered/common/mixin/api/minecraft/world/entity/EntityMixin_API.java [hunk 2]
```diff
@@ -167,7 +172,7 @@ public boolean setTransform(final Transform transform) {
         }
         Preconditions.checkNotNull(transform, "The transform cannot be null!");
         final Vector3d position = transform.position();
-        this.shadow$setPos(position.x(), position.y(), position.z());
+        this.setPosition(transform.position());
```
Comment (Zidane) [inline src/mixins/java/org/spongepowered/common/mixin/api/minecraft/world/entity/EntityMixin_API.java]: Do not make API calls, the below code shouldn't either IF there is a bridge method.

File: src/mixins/java/org/spongepowered/common/mixin/core/world/entity/EntityMixin.java [hunk 3]
```diff
@@ -240,6 +241,32 @@
...
+            Vector3d destinationPosition = position;
+            if (ShouldFire.MOVE_ENTITY_EVENT) {
+                final MoveEntityEvent event = SpongeEventFactory.createMoveEntityEvent(frame.currentCause(),
+                    (org.spongepowered.api.entity.Entity) this, VecHelper.toVector3d(this.shadow$position()), position, position);
+                if (SpongeCommon.postEvent(event)) {
+                    return false;
+                }
+                destinationPosition = event.destinationPosition();
+            }
+            this.shadow$teleportToWithTicket(destinationPosition.x(), destinationPosition.y(), destinationPosition.z());
```
Comment (dualspiral) [inline src/mixins/java/org/spongepowered/common/mixin/core/world/entity/EntityMixin.java:261]: `teleportToWithTicket` has been overwritten to fire the event anyway. Shouldn't you just call that and not fire the event here?

File: src/mixins/java/org/spongepowered/common/mixin/core/world/entity/EntityMixin.java [hunk 4]
```diff
@@ -265,52 +293,42 @@
...
-            this.bridge$revive();
-            this.shadow$setLevel(destinationWorld);
-            destinationWorld.addFromAnotherDimension((Entity) (Object) this);
-
-            originalWorld.resetEmptyTime();
-            destinationWorld.resetEmptyTime();
-
-            final ChunkPos chunkPos = new ChunkPos((int) this.shadow$getX() >> 4, (int) this.shadow$getZ() >> 4);
-            destinationWorld.getChunkSource().addRegionTicket(TicketType.POST_TELEPORT, chunkPos, 1, ((Entity) (Object) this).getId());
+            this.shadow$teleportToWithTicket(destinationPosition.x(), destinationPosition.y(), destinationPosition.z());
```
Comment (dualspiral) [inline src/mixins/java/org/spongepowered/common/mixin/core/world/entity/EntityMixin.java:328]: Same problem again here - no need to fire the event above if it's going to happen in this method.

**Review by dualspiral — COMMENTED** [review]
As well as the comments below (you're firing an event twice), you'll need to have different handling for `ServerPlayer` because of the way it works with packets. See how it's done in `TeleportCommandMixin` - you may want to unify the logic a little.

**Comment by dualspiral** [comment]
#3484 - but thanks for this, it really helped me get this to where it is now!

**Pull request closed without merging**